Scheduler.lib.nut

```diff
@@ -0,0 +1,423 @@
+        if (_jobs.len() > 0) {
+            if (_currentJob != null) imp.cancelwakeup(_currentJob);
+
+            local dur = _getDur(_jobs[0]);
+
+            _currentJob = imp.wakeup(dur, function() {
+                _jobs[0].cb.acall(_jobs[0].args);
+                _next();
+            }.bindenv(this));
+            _currentId = _jobs[0].id;
+        }
+    }
+
+    // Cancel an existing timer
+    //
+    // Parameters:
+    //     id (integer)       the id of the existing timer
+    // Return: (boolean) whether the timer was removed or not
+    function _cancel(id) {
+        // If the timer to cancel is currently running, cancel it
+        if (_currentId == id) {
+            imp.cancelwakeup(_currentJob);
+            _currentJob = null;
+            _currentId = null;
+        }
+
+        // Look for the timer in the queue and remove it
+        foreach (i, t in _jobs) {
+            if (_jobs[i].id == id) {
+                _jobs.remove(i);
+                // Check if the timer to cancel is at the front of the queue
+                if (i == 0 && _jobs.len() >= 1) {
+                    _next();
+                }
+            }
+        }
+    }
+
+    // Trigger an existing timer's callback to run now (timer will also continue as normal)
+    //
+    // Parameters:
+    //     id (integer)       the id of the existing timer
+    // Return: (boolean) whether the callback was triggered or not
+    function _now(id) {
+        foreach (i, t in _jobs) {
+            if (_jobs[i].id == id) {
+                _jobs[i].cb.acall(_jobs[i].args);
+            }
+        }
+    }
+
+    // Get the current time differently on device and agent
+    function _getTime() {
+        if (_env == ENVIRONMENT_AGENT) return date();
+        else                           return hardware.millis() / 1000.0;
+    }
+
+    // Calculate "second" count for a timer differently on device and agent
+    function _getSec(dur, now) {
+        if (_env == ENVIRONMENT_AGENT) return math.floor(dur).tointeger() + now.time;
+        else                           return math.floor(dur).tointeger() + math.floor(now).tointeger();
+    }
+
+    // Calculate "sub second" count for a timer differently on device and agent
+    function _getSubSec(dur, now) {
+        if (_env == ENVIRONMENT_AGENT) return dur - math.floor(dur).tointeger() + (now.usec / 1000000.0);
+        else                           return dur - math.floor(dur).tointeger() + (now - math.floor(now).tointeger());
+    }
+
+    function _getDur(job) {
+        if (_env == ENVIRONMENT_AGENT) {
+            local now = date();
+            return (job.sec - now.time) + (job.subSec - now.usec / 1000000.0);
+        } else {
+            local now = hardware.millis() / 1000.0;
+            local nowSec = math.floor(now).tointeger();
+            local nowSubSec = (now) - math.floor(now).tointeger();
+
+            return (job.sec - nowSec) + (job.subSec - nowSubSec);
+        }
+    }
+
+}
+
+class Job {
```

Comment:
I would move this class inside `Scheduler`. I.e.

```
class Scheduler {
    Job = class {
    }
}
```

Reply:
The reasoning here is that the Job only makes sense in the context of the Scheduler class.

Reply:
I would rather do it like this just for readability if it's ok with you?:
```
class Scheduler.Job {
}
```

Reply:
Either of these formats are fine, but the current code should be updated so Job class is not just Job.